frontend/catalyst/preprocess.py

```diff
@@ -16,16 +16,120 @@
 import jax
 import pennylane as qml
 from pennylane import transform
-from pennylane.measurements import CountsMP, ExpectationMP, ProbabilityMP, VarianceMP
+from pennylane.measurements import (
+    CountsMP,
+    ExpectationMP,
+    MidMeasureMP,
+    ProbabilityMP,
+    VarianceMP,
+)
 from pennylane.tape.tape import (
     _validate_computational_basis_sampling,
     rotations_and_diagonal_measurements,
 )
 
 import catalyst
+import catalyst.pennylane_extensions
+from catalyst.tracing.contexts import EvaluationContext
 from catalyst.utils.exceptions import CompileError
 
 
+def _operator_decomposition_gen(
+    op: qml.operation.Operator,
+    acceptance_function,
+    decomposer,
+    max_expansion=None,
+    current_depth=0,
+):
+    """A generator that yields the next operation that is accepted."""
+    max_depth_reached = False
+    if max_expansion is not None and max_expansion <= current_depth:  # pragma: no cover
+        max_depth_reached = True
+    if acceptance_function(op) or max_depth_reached:
+        yield op
+    else:
+        try:
+            decomp = decomposer(op)
+            current_depth += 1
+        except qml.operation.DecompositionUndefinedError as e:  # pragma: no cover
+            raise CompileError(
+                f"Operator {op} not supported on device and does not provide a decomposition."
+            ) from e
+
+        for sub_op in decomp:
+            yield from _operator_decomposition_gen(
+                sub_op,
+                acceptance_function,
+                decomposer=decomposer,
+                max_expansion=max_expansion,
+                current_depth=current_depth,
+            )
+
+
+@transform
+def decompose(
+    tape: qml.tape.QuantumTape,
+    ctx,
+    stopping_condition,
+    max_expansion=None,
+):
+    """Decompose operations until the stopping condition is met."""
+
+    def decomposer(op):
+        if op.name in {"MultiControlledX", "BlockEncode"} or isinstance(op, qml.ops.Controlled):
+            try:
+                mat = op.matrix()
+            except Exception as e:
+                raise CompileError(
+                    f"Operation {op} could not be decomposed, it might be unsupported."
+                ) from e
+            op = qml.QubitUnitary(mat, wires=op.wires)
+            return [op]
+        elif isinstance(
+            op,
+            (
+                catalyst.pennylane_extensions.Adjoint,
+                catalyst.pennylane_extensions.MidCircuitMeasure,
+                catalyst.pennylane_extensions.ForLoop,
+                catalyst.pennylane_extensions.WhileLoop,
+                catalyst.pennylane_extensions.Cond,
+            ),
+        ):
+            for region in op.regions:
+                if region.quantum_tape:
+                    with EvaluationContext.frame_tracing_context(ctx, region.trace):
```

Comment:
@lillian542 Notice here I add to add the right Jax fram otherwise we have leaked values issues.

Reply:
Where does the jax frame come from when calling `decompose`?

Reply:
There is a trace associated to each region for the hybrid op, `with EvaluationContext.frame_tracing_context(ctx, region.trace):` gives the right scope for the decomposition to happen. Otherwise it considers that the tracer are leaked when in the decomposition you actually uses the parameters: e.g. `op -> [RX(phi / 8, wires=0), RX(phi / 8, wires=0)]`